casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -141,6 +151,49 @@ object Validate {
       }
     }
 
+  /**
+    * Validate no deploy by the same (user, millisecond timestamp)
+    * has been produced in the chain
+    */
+  def repeatDeploy[F[_]: Monad: Log: BlockStore](block: BlockMessage,
+                                                 genesis: BlockMessage,
+                                                 dag: BlockDag): F[Boolean] =
+    BlockStore[F].asMap().flatMap { internalMap: Map[BlockHash, BlockMessage] =>
+      {
+        def parents(b: BlockMessage): Iterator[BlockMessage] =
+          ProtoUtil.parents(b).iterator.flatMap(internalMap.get)
+
+        val deployKeySet = block.getBody.newCode.map(d => (d.getRaw.user, d.getRaw.timestamp)).toSet
```

Comment:
I'm a little worried about these `get*` methods. I know that the technically optional fields they get should always be present, but we don't want to make it trivially easy to crash the node by simply sending it an empty message. I would prefer if you used a for-comprehension like
```scala
(for {
  bd <- block.body
  d  <- bd.newCode
  r  <- d.raw
} yield (r.user, r.term, r.timestamp)).toSet
```

Reply:
Done.